### ISSUE TITLE
bugfinds and other accuracy issues

### ISSUE THREAD

**Issue by robinerd123** [gonetz/GLideN64#1899]
so i have a few problems with GLideN64, using that as a main gfx plugin for testing my homebrew i never realized about a lot of issues that were there and stuff, so here are a few things

- texturerectangle wrapping not emulated correctly
- it ignoring the cycletype in some cases
- alpha combiner not emulated correctly (texel0 and texel1 are switched or something similar to that)
- loadtile working with less things set than what should actually be set to make it work properly

and i also had a few other bugs/issues but i cant manage to replicate them lol, ill keep trying to replicate them and get to update you with the stuff

**Comment by StenApp** [comment]
Which emulator, which version, which graphics card, which game is affected?

**Comment by olivieryuyu** [comment]
@robinerd123 

please fell free to submit commits then.

**Comment by robinerd123** [comment]
im mainly using pj64 dev but i get the same results in mupen64plus, and the game is my homebrew, i cant provide screenshots with differences and what the problems particularly are now because the internet here is terrible but ill put up some reference screenshots asap

**Comment by robinerd123** [comment]
ok ill do a quick example now of the alpha combiner not being emulated correctly,
https://m.imgur.com/gallery/5LpAhAb

**Comment by robinerd123** [comment]
yes

**Comment by gonetz** [comment]
@robinerd123 Could you build test roms, which I can use to debug the issue you  reported? As few polygons as possible, just to see the problem. I can't fix issues just by their description.

**Comment by robinerd123** [comment]
https://ufile.io/on7gt
this rom is fine in angrylion but in GLideN64 the alpha combiner is messed up, its just the same as the screenshot i sent, i already patched some other bugs that didnt allow it to show well in angrylion but ill replicate them and upload other roms soon

**Comment by gonetz** [comment]
@binerd123 You do weird things with combiners in your homebrew.
angrylion explained somewhere, how RDP treats tiles in 2cycles mode.
At first cycle it uses tile0 for TEXEL0 input of combiner and tile1 for TEXEL1 input.
At second cycle it uses tile1 as TEXEL0 input and tile0 + s_offset as  TEXEL1 input.
That is tiles are swapped indeed on 2nd cycle, but for tile0 it takes next texel.
GLideN64 does not emulate tile shift in second cycle, see issue #25

Now, you use 2 texrects in 2cycle mode for background of GALAXY N64 logo.
First texrects uses combiner:
1st cycle:
color (0 - 0) *TEXEL0 + TEXEL0
alpha (0-0) * 0 + TEXEL0
2nd cycle:
color (1 -TEXEL0) * COMBINED + PRIMITIVE
alpha (0-0) * 0 + TEXEL0

Thus, color after first cycle is TEXEL0, alpha is TEXEL0 too. That is color and alpha are fetched from tile 0.
Substitute it to color and alpha equations for the 2nd cycle:
color: (1 - color_of_tile1) * color_of_tile0 + prim
alpha alpha_of_tile1
Are you sure it is what you want?
Did you set tile1 properly?
Also, pay attention on blending mode, as you set FORCE_BL enabled.
You use CLR_IN * A_IN + CLR_MEM * 1MA for both cycles.

The second texrect is even weirder. 
1st cycle:
color (0 - 0) *0 + PRIMITIVE
alpha (1-TEXEL1) * TEXEL0 + 0
2nd cycle is the same:
color (0 - 0) *0 + PRIMITIVE
alpha (1-TEXEL1) * TEXEL0 + 0

That is result of first cycle is ignored. As I said, second cycle of combiner uses tile1 as TEXEL0 input and tile0 + s_offset as  TEXEL1 input. As I see, you use texture of width 1 as tile0. I'm not sure what should be fetched from it after offset is applied. May be zero. Anyway, texture shift is not implemented in GLideN64, and in this case it can be the source of the problem.

**Comment by theboy181** [comment]
Test ROM is buggy for me in AL+ is this to be expected?

**Comment by olivieryuyu** [comment]
Now does it look on real n64?

**Comment by theboy181** [comment]
Ill test after work, good point though.

**Comment by robinerd123** [comment]
yeah @gotenz i did alot of weird stuff there with the combiner but it seemed to be  fine with angrylion so meh, and @theboy181 that rom that i sent's background behind the logo crashes on console due to a issue i didnt fix yet so if you want to test that on console you need a rom that uses a different background model, which you can try https://ufile.io/byqjm

and @gotenz again here is a rom

**Comment by robinerd123** [comment]
yeah @gotenz i did alot of weird stuff there with the combiner but it seemed to be  fine with angrylion so meh, and @theboy181 that rom that i sent's background behind the logo crashes on console due to a issue i didnt fix yet so if you want to test that on console you need a rom that uses a different background model, which you can try https://ufile.io/byqjm

and @gotenz again here is a rom demonstrating texture rectangles not working properly, here it works 100% fine in GLideN64 but it has issues in angrylion, cen, and also on console, its in the game select screen that you can access by pressing start, and you can see it in this old rom from before i fixed the issue with some kinda cheap method https://ufile.io/ihum3

**Comment by gonetz** [comment]
@robinerd123 Can you make texture for the second texrect with 2 texel width and check how it will work with angrylion's plugin and with GLideN64?

**Comment by robinerd123** [comment]
oops, pressed wrong button lol
you mean the second textrect in the galaxy64 text? and do you want me to rescale the gradient texture to be 2x60 too? actually i think that might not fit in tmem because im using that other huge ia4 texture for the mask

**Comment by gonetz** [comment]
Yes, can you double width of the gradient texture? Will it change result with angrylion's plugin?
